### PR TITLE
[AC-607] Extract IOrganizationService.DeleteUserAsync into IRemoveOrganizationUserCommand

### PR DESCRIPTION
**Pull request by r-tome**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/AC-607

## 📔 Objective

> Move logic from `IOrganizationService.RemoveUserAsync` to `IRemoveOrganizationUserCommand`, remove `IOrganizationService.RemoveUserAsync` and update all dependencies to use `IRemoveOrganizationUserCommand`.
> 
> Move related tests from `OrganizationServiceTests` to  `RemoveOrganizationUserCommandTests`.

I have also extracted the method `HasConfirmedOwnersExceptAsync` from `OrganizationService` into its own query because otherwise there would be a circular dependency between `OrganizationService` and `RemoveOrganizationUserCommand`.

I also took the opportunity to update a OrganizationUsersController and OrganizationsController and refactor controller actions to accept `Guid` directly instead of parsing strings.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/RemoveOrganizationUserCommand.cs [hunk 1]
```diff
@@ -8,38 +10,171 @@ namespace Bit.Core.AdminConsole.OrganizationFeatures.OrganizationUsers;
...
+
+        await _organizationUserRepository.DeleteAsync(orgUser);
+
+        if (orgUser.UserId.HasValue)
+        {
+            await DeleteAndPushUserRegistrationAsync(orgUser.OrganizationId, orgUser.UserId.Value);
+        }
+    }
+
+    private async Task<IEnumerable<KeyValuePair<string, DeviceType>>> GetUserDeviceIdsAsync(Guid userId)
```
Comment (r-tome) [inline src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/RemoveOrganizationUserCommand.cs:165]: I have copied these methods from `OrganizationService`. Should I create a new command for this? Or is too much scope creep at this point?

**Review by cturnbull-bitwarden — DISMISSED** [review]

File: test/Core.Test/AdminConsole/OrganizationFeatures/OrganizationUsers/HasConfirmedOwnersExceptQueryTests.cs
Comment (JaredSnider-Bitwarden) [inline test/Core.Test/AdminConsole/OrganizationFeatures/OrganizationUsers/HasConfirmedOwnersExceptQueryTests.cs:1]: Love to see all the tests!

**Review by JaredSnider-Bitwarden — APPROVED** [review]
Really nice clean up effort across the board! Auth files look great! A single non-blocking feedback item.

**Review by jonashendrickx — DISMISSED** [review]

**Review by jrmccannon — DISMISSED** [review]

File: src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/RemoveOrganizationUserCommand.cs (hunk 1 above)
Comment (jrmccannon) [reply src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/RemoveOrganizationUserCommand.cs:165]: Probably too much at this point.

File: src/Api/AdminConsole/Controllers/OrganizationsController.cs [hunk 2]
```diff
@@ -74,7 +76,8 @@ public OrganizationsController(
         IPushNotificationService pushNotificationService,
         IProviderRepository providerRepository,
         IProviderBillingService providerBillingService,
-        IDataProtectorTokenFactory<OrgDeleteTokenable> orgDeleteTokenDataFactory)
+        IDataProtectorTokenFactory<OrgDeleteTokenable> orgDeleteTokenDataFactory,
+        IRemoveOrganizationUserCommand removeOrganizationUserCommand)
```
Comment (jonashendrickx) [inline src/Api/AdminConsole/Controllers/OrganizationsController.cs:80]: @withinfocus & @JaredSnider-Bitwarden I think it would be better to switch to action injection instead of constructor injection. We could save on memory consumption for every incoming request.

https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/dependency-injection?view=aspnetcore-8.0#action-injection-with-fromservices

You can inject it directly.

```csharp
public async Task Leave(Guid id, [FromServices] IRemoveOrganizationUserCommand removeOrganizationUserCommand)
```

That's 32 bytes saved at minimum for every other action in `OrganizationsController`

File: src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/Interfaces/IHasConfirmedOwnersExceptQuery.cs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Bit.Core.AdminConsole.OrganizationFeatures.OrganizationUsers.Interfaces;
+
+public interface IHasConfirmedOwnersExceptQuery
+{
+    Task<bool> HasConfirmedOwnersExceptAsync(Guid organizationId, IEnumerable<Guid> organizationUsersId, bool includeProvider = true);
```
[the remark this answers is not in the document]
Comment (r-tome) [reply src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/Interfaces/IHasConfirmedOwnersExceptQuery.cs:11]: I don't have a better name for it (let me know if you do) but I added a `summary` to the interface

File: src/Api/AdminConsole/Controllers/OrganizationsController.cs (hunk 2 above)
Comment (r-tome) [reply src/Api/AdminConsole/Controllers/OrganizationsController.cs:80]: @jonashendrickx , that's pretty neat! I wasn't aware of that, thanks! I'll wait for @withinfocus  to decide whether we want to use it, as it breaks our usual pattern.

Comment (justindbaur) [reply src/Api/AdminConsole/Controllers/OrganizationsController.cs:80]: My opinion on the matter:

For controllers, it's mostly stylistic, @jonashendrickx is right there are a small amount of memory savings for all other actions if the service is transient or scoped and hasn't already been resolved. But I would never make this choice for those savings. If you care that much, don't use a controller action to begin with, start using minimal API's. But then speaking of minimal API's, they are likely a thing that will get used more and more in Bitwarden code and they will use method injected services like this so it's not a bad idea to get people used to it.

**Review by jrmccannon — APPROVED** [review]

File: src/Api/AdminConsole/Controllers/OrganizationsController.cs (hunk 2 above)
Comment (r-tome) [reply src/Api/AdminConsole/Controllers/OrganizationsController.cs:80]: Thanks for the input. I'll leave this as is for now to keep things consistent with our current pattern. We can revisit this if we move towards minimal APIs more broadly.

**Review by jonashendrickx — APPROVED** [review]

**Review by JaredSnider-Bitwarden — APPROVED** [review]
Auth changes LGTM!
